Make app_settings file more modular

njogz opened the pull request:
This PR does the following:

- Moves manually defined configurations from the app_settings.json file so that that can be generated by medic-conf and can be ignored by git to help avoid merge conflicts when working on configurations.
- Moves SMS schedule information and SMS form definitions to separate files similar to how tasks are defined.

[214](https://github.com/medic/medic-conf/issues/214)
[68](https://github.com/medic/medic-conf/issues/68)

dianabarsan commented on src/fn/compile-app-settings.js:
```diff
@@ -39,7 +39,30 @@ const compileAppSettingsForProject = async (projectDir, options) => {
-  const appSettings = fs.readJson(path.join(projectDir, 'app_settings.json'));
+  // Manual configurations should be done in the base-settings.json file
+  // This check and warning can be removed when all project configurations have this new file defined
+  let appSettings = fs.readJson(path.join(projectDir, 'app_settings.json'));
```
This line will thrown an error if the app_settings file does not exist. Does this mean we expect configurers to create this file and leave it empty if they wish to go the modular route?

njogz replied:
Added this so as not to break current configurations. Will refactor based on the suggestion below on checking if config is modular.

dianabarsan commented on test/data/compile-app-settings/base-settings/project/targets.json:
```diff
@@ -0,0 +1 @@
+{}
```
what is the purpose of this file?

dianabarsan commented on src/fn/compile-app-settings.js:
```diff
@@ -39,7 +39,30 @@ const compileAppSettingsForProject = async (projectDir, options) => {
+  const formSettings = readOptionalJson(path.join(projectDir, 'app_settings/forms.json'));
+  if (formSettings) {
+    if (appSettings.forms && Object.keys(appSettings.forms).length !== 0) {
+      throw new Error(`forms is defined in both the base settings and the forms.json file.`);
```
I'm not sure what to say about this. If the project doesn't have `base_settings.js`, but has `app_settings` and `forms.js`, then the user will get an error every time they try to generate config. 

I think we should gate all these errors in one block, where we test whether app_settings are indeed modularised, and only use these files if they are.

njogz replied:
The idea is not to have two places that might have conflicting config of the same thing. I like the idea of having one check for modular config.

jonathanbataire commented:
Hey @njogz thanks for this
did some feedback:
- I think we should incorporated this project structure when initializing project layout
- should we consider automating restructuring for old config or we have to manually edit app-settings (maybe I missed sth)
- When i put everything in base_settings (including forms and schedules) everything compiles fine with no warnings 
i was hoping forms and schedules would throw a warning when not put in the schedules.json and forms.json files 
- Do we do any validation on the content of these files(base_settings,schedules and forms) looks like anything can go anywhere

Will continue to test.............

njogz commented:
Hi @jonathanbataire. Thank you for the thoughtful feedback. See my response below:

- Added the new structure the initialisation command
- We are keeping the restructuring to be manual for now
- I have added warnings if either forms or schedules is defined in base_settings
- Validation is brilliant idea and added some validation for the forms.json and schedules.json files

jonathanbataire commented:
Hey @njogz So i noticed that when i initialize-project-layout `schedules.json` is created as an object, this is supposed to be an array
Otherwise everything else i pointed out works well now. Thanks......

dianabarsan commented on src/fn/compile-app-settings.js:
```diff
@@ -39,7 +40,43 @@ const compileAppSettingsForProject = async (projectDir, options) => {
+  let appSettings;
+  const baseSettingsPath = path.join(projectDir, 'app_settings/base_settings.json');
+  const appSettingsPath = path.join(projectDir, 'app_settings.json');
+  if (!fs.exists(baseSettingsPath) && !fs.exists(appSettingsPath)) {
```
this is a nit, but I'd add some empty lines between logical groups to improve readability.

dianabarsan commented on src/lib/validate-app-settings.js:
```diff
@@ -0,0 +1,74 @@
+        position: joi.number(),
+        flags: joi.object(),
+        length: joi.array().items(joi.number()),
+        required: joi.boolean()
```
I've run this over standard config and it fails because some forms have fields that have `validations` and `range` properties. 
I'm not sure if these fields are even used, but we should definitely make sure. 

I think we should run this against a number of SMS projects configurations and see if they pass.

dianabarsan reviewed (changes requested):
Looking good! I left a couple of minor comments about code styling and one comment about schema validations.

dianabarsan reviewed (approved):
Looks good!
